## Auralin Phase II clinical trial - Data Wrangle
The Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. 
This project focuses on **data wrangling** in order to compare the efficacy and safety of a new oral insulin to treat diabetes to injectable insulin. 

### Table of Contents

- <a href='#intro'>Introduction</a>
- <a href='#datawrangling'>Data Wrangling</a>
    - <a href='#gather'>Gather</a>
    - <a href='#assess'>Assess</a>
    - <a href='#clean'>Clean</a>
- <a href='#conclusion'>Conclusion</a>   

<a id='intro'></a>
# Introduction

> **DISCLAIMER: This Data Isn't "Real"**
The Auralin and Novodra are not real insulin products. This clinical trial data was fabricated for the sake of this project. When assessing this data, the issues that you'll detect (and later clean) are meant to simulate real-world data quality and tidiness issues.

> That said:
> - This dataset was constructed with the consultation of real doctors to ensure plausibility by Udacity.
> - This clinical trial data for an alternative insulin was inspired and closely mimics this real [clinical trial for a new inhaled insulin called Afrezza](https://care.diabetesjournals.org/content/38/12/2266.long).
> - The data quality issues in this dataset mimic real, [common data quality issues in healthcare data](http://media.hypersites.com/clients/1446/filemanager/Articles/DocCenter_Problem_with_data.pdf). These issues impact quality of care, patient registration, and revenue.
> - The patients in this dataset were created using this [fake name generator](https://www.fakenamegenerator.com/order.php) and do not include real names, addresses, phone numbers, emails, etc.

<a id='datawrangling'></a>
## Data Wrangling

In [1]:
# Import required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

<a id='gather'></a>
## Gather

> Data has been already created by Udacity and readily available for further processing

In [2]:
# read from the data files and store it in respective dataframe

patients = pd.read_csv('data/patients.csv')
treatments = pd.read_csv('data/treatments.csv')
adverse_reactions = pd.read_csv('data/adverse_reactions.csv')

<a id='assess'></a>
## Assess

> **Types of assessment:**
> - <a href='#visual'>Visual</a>
> - <a href='#prog'>Programmatic</a>

<a id='visual'></a>
### Visual assessment

In [6]:
# Display the patients table
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [4]:
# Display the treatments table
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [5]:
# Display the adverse_reactions table
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

<a id='prog'></a>
### Programmatic assessment

In [22]:
# info on patients table
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [25]:
patients[patients.address.isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [23]:
# info on treatments table
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [24]:
# info on adverse_reactions table
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [26]:
# describe on patients
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [27]:
# describe on treatments
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [29]:
# there is huge difference between 50% and 75% of hba1c_change, although the difference in hba1c_start and hba1c_end is less.
# let's look at the data containing higher values
treatments[treatments.hba1c_change > 0.50]

given_name     surname auralin    novodra  hba1c_start  hba1c_end  \
1       elliot  richardson       -  40u - 45u         7.56       7.09   
13      gregor        bole       -  47u - 45u         7.61       7.16   
17        gina        cain       -  36u - 36u         7.88       7.40   
20        mile      stanić       -  47u - 48u         7.66       7.24   
25      benoît      bonami       -  44u - 43u         9.82       9.40   
..         ...         ...     ...        ...          ...        ...   
245         wu        sung       -  47u - 48u         7.61       7.12   
257   mathilde    nørgaard       -  27u - 28u         8.50       8.10   
267       leon      scholz       -  38u - 32u         7.72       7.29   
268     yasmin      araujo       -  51u - 54u         7.82       7.36   
272     steven         roy       -  41u - 43u         7.87       7.43   

     hba1c_change  
1            0.97  
13           0.95  
17           0.98  
20           0.92  
25           0.92  
..            ...  
245          0.99  
257          0.90  
267          0.93  
268          0.96  
272          0.94  

[61 rows x 7 columns]

**Note :** hba1c_change = hba1c_start - hba1c_end
> - From the data it is clear that the value is computed inaccurately

In [49]:
# sample on patients table 
patients.sample(5)

patient_id assigned_sex given_name  surname                 address  \
126         127         male       Mads  Nielsen    1493 Randolph Street   
413         414         male        Kai    Cunha    2967 Prudence Street   
403         404         male     Robert   Maslov        2356 Myra Street   
328         329       female       Anja   Hueber    3216 Lodgeville Road   
407         408       female      Tegan  Johnson  2636 Benson Park Drive   

              city state  zip_code        country  \
126         Boston    MA    2210.0  United States   
413        Detroit    MI   48221.0  United States   
403     Providence    RI    2908.0  United States   
328    Minneapolis    MN   55402.0  United States   
407  Oklahoma City    OK   73107.0  United States   

                                     contact  birthdate  weight  height   bmi  
126  MadsANielsen@jourrapide.com508-739-5632  5/19/1942   137.1      68  20.8  
413    313-341-7799KaiOliveiraCunha@cuvox.de   2/1/1990   163.5      74  21.0  
403     RobertMaslov@fleckens.hu401-535-2675   7/2/1932   219.1      65  36.5  
328  AnjaHueber@teleworm.us+1 (612) 342-6065  4/16/1987   151.8      65  25.3  
407       405-470-5088TeganJohnson@gustr.com  1/19/1928   156.6      65  26.1

In [51]:
patients.surname.value_counts()

Doe           6
Jakobsen      3
Taylor        3
Berg          2
Tucker        2
             ..
Vũ            1
Martinsson    1
Marchesi      1
Citizen       1
Reilly        1
Name: surname, Length: 466, dtype: int64

In [53]:
patients.address.value_counts()

123 Main Street               6
2476 Fulton Street            2
2778 North Avenue             2
648 Old Dear Lane             2
2322 Grim Avenue              1
                             ..
2687 Black Oak Hollow Road    1
2126 Pearl Street             1
2884 Elsie Drive              1
1557 Byers Lane               1
465 Southern Street           1
Name: address, Length: 483, dtype: int64

In [54]:
patients[patients.address.duplicated(keep=False)]

patient_id assigned_sex given_name      surname             address  \
24           25         male      Jakob     Jakobsen   648 Old Dear Lane   
29           30         male       Jake     Jakobsen   648 Old Dear Lane   
97           98         male    Patrick      Gersten   2778 North Avenue   
131         132       female     Sandra       Taylor  2476 Fulton Street   
209         210       female     Lalita  Eldarkhanov                 NaN   
215         216         male       John          Doe     123 Main Street   
219         220         male         Mỹ        Quynh                 NaN   
229         230         male       John          Doe     123 Main Street   
230         231       female  Elisabeth      Knudsen                 NaN   
234         235       female    Martina    Tománková                 NaN   
237         238         male       John          Doe     123 Main Street   
242         243         male       John      O'Brian                 NaN   
244         245         male       John          Doe     123 Main Street   
249         250         male   Benjamin       Mehler                 NaN   
251         252         male       John          Doe     123 Main Street   
257         258         male        Jin         Kung                 NaN   
264         265       female   Wafiyyah       Asfour                 NaN   
269         270       female     Flavia   Fiorentino                 NaN   
277         278         male       John          Doe     123 Main Street   
278         279       female   Generosa        Cabán                 NaN   
282         283       female      Sandy       Taylor  2476 Fulton Street   
286         287         male      Lewis         Webb                 NaN   
296         297       female        Chỉ          Lâm                 NaN   
502         503         male        Pat      Gersten   2778 North Avenue   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
97          Burr        NE   68324.0  United States   
131     Rainelle        WV   25962.0  United States   
209          NaN       NaN       NaN            NaN   
215     New York        NY   12345.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
97         PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   
131        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
209                                         NaN   8/14/1950   143.4      62   
215                 johndoe@email.com1234567890    1/1/1975   180.0      72   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230              

In [59]:
# weight of the patients
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [62]:
# Let's check whether the 48.8 is in kg by converting it into lbs and calculating its bmi
# and check with the actual data
weight_lbs = patients.iloc[210].weight * 2.20462
height_in = patients.iloc[210].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

19.05582654774502

In [63]:
patients.iloc[210].bmi

19.1

**Note :** Both are close before rounding

In [58]:
treatments.auralin.isna().sum() , treatments.novodra.isna().sum()

(0, 0)

In [65]:
# Duplicate column names 
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [66]:
# given_name and surname column names are duplicated 
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

### Quality issues

#####  `patients` table
- zip code is a float not a string
- zip code has four digits sometimes (must be 5 digits) 
- Tim Neudorf height is 27 in instead of 72
- full state name sometimes, abbreviations other times
- given_name in id=9, mispelled
- Missing demographic information for few rows (address - contact colummns)
- Erroneous datatypes (assigned sex, state, zip_code and birthdate columns)
- multiple phone number formats
- default duplicates (John Doe) has replaced the real data (real data missing)
- Multiple records for Jakobsen, Taylor, Gersten
- Lowest weight 48.8 in kg instead of lbs

#####  `treatments` table
- missing values for 'hba1c_change'
- inaccurate Hba1c changes
- 'u' next to the  start and end dose of the <i>auralin</i> and <i>novodra</i> columns
- missing records (280 instead of 350)
- Lowercase given names & surnames
- Erroneous datatypes (auralin and novodra columns)
- Nulls represented as dashes (-) in auralin and novodra columns

#####  `adverse_reactions` table
- Lowercase given names & surnames

### Tidiness issues

#####  `patients` table
- contacts columns contains both phone_number and email details

#####  `treatments` table
- auralin and novodra are values, not treatment variable
- both auralin and novodra contains two variable details, start dose and end dose
- patient_id : missing unique identifier
- duplicate column names in both treatment & adverse_reactions tables (given_name, surname)

<a id='clean'></a>
## Clean

In [67]:
# Make a copy of three dataframes
patients_copy = patients.copy()
treatments_copy = treatments.copy()
adverse_reactions_copy = adverse_reactions.copy()

<a id='conclusion'></a>
## Conclusion